---
layout: home
permalink: index.html

# Please update this with your repository name and title
repository-name: eYY-4yp-project-template
title:
---

[comment]: # "This is the standard layout for the project, but you can clean this and use your own template"

# Project Title

#### Team

- E/17/134, Kavindu Jayasooriya, [email](mailto:e17134@eng.pdn.ac.lk)
- E/17/207, Pasindu Marasinghe, [email](mailto:e17207@eng.pdn.ac.lk)
- E/17/318, Udith Senanayake, [email](mailto:e17318@eng.pdn.ac.lk)

#### Supervisors

- Prof. Roshan G. Ragel, [email](mailto:name@eng.pdn.ac.lk)
- Dr. Hasindu Gamaarachchi, [email](mailto:name@eng.pdn.ac.lk)

#### Table of content

1. [Abstract](#abstract)
2. [Related works](#related-works)
3. [Methodology](#methodology)
4. [Experiment Setup and Implementation](#experiment-setup-and-implementation)
5. [Results and Analysis](#results-and-analysis)
6. [Conclusion](#conclusion)
7. [Publications](#publications)
8. [Links](#links)

---

## Abstract

In this research, our aim is to explore various approaches to enhance the compression ratios of raw nanopore data. The majority of extant literature in this domain has primarily focused on lossless compression techniques. However, we posit that this realm has reached a saturation point, and therefore, there exists a greater potential for improvement in the lossy compression domain. Our research is centred on identifying the noise present in the signal and its consequential impact on downstream applications.

## Introduction

DNA sequencing proves to be a formidable instrument in acquiring data concerning the genetic composition of living entities.A Genome encompasses the complete set of requisite data to construct and preserve the living entity, comprising its corporeal attributes, its operational efficiency, and its reproductive mechanisms.

The genome comprises DNA (or RNA in RNA viruses), a lengthy molecule composed of nucleotides. Identifying the sequential arrangement of the four nucleotides, namely 'adenine (A), guanine (G), thymine (T), and cytosine (C),' within a DNA molecule is commonly referred to as DNA sequencing. This methodology is employed to observe the genome and the proteins they encode of a given species or specific fragments of DNA, which is advantageous in identifying various genetic variations caused by diseases and potentially creating novel medicines to address them.

Nanopore Sequencing, a third-generation sequencing methodology, involves the recording of the ionic current that occurs as a DNA molecule transits through a nano-scale protein pore, also referred to as a nanopore. The vast quantities of data produced by Nanopore Sequencing necessitate adherence to certain data regulations which mandate that the sequenced data be archived for an extended duration (several years). However, as the rate of Nanopore Sequencing increases, the cost of data storage also rises exponentially. To provide an example, a compressed state-of-the-art dataset for a single human genome generated by Nanopore Sequencing is approximately 1 TB, and retaining this data for numerous years can prove to be excessively expensive. The Garvan Institute of Medical Research generates DNA sequencing data which must be archived for a period of up to a decade, and the annual data volume ranges between 260 and 520 TBs, resulting in a staggering USD 6000 expenditure to preserve the data in Google Cloud Storage.

The procedure of deducing the accurate nucleotides from the unprocessed signal produced by the DNA sequencer is called base-calling. This field is highly dynamic and ever-evolving, with novel techniques being constantly developed to improve the accuracy and efficacy of base-calling. Therefore, it is crucial to retain the raw data in storage rather than archiving the base-called outcomes, since the raw data can prove to be instrumental in base-calling research.

When it comes to the objective of minimizing archiving costs, it is of utmost importance to engage in the compression of nanopore raw signal data. In the context of minimizing archiving costs, it is crucial to engage in the compression of nanopore raw signal data. Therefore, it is necessary to archive the raw data and explore innovative compression methodologies to deliver superior and more cost-effective solutions.

## Related works

Very few studies focus on the compression of nanopore signal data, even though there are many research studies related to genomic data compression. Gzip is a lossless compression method for compressing different data types, such as text and images. A tool called Picopore \cite{Gigante2017} increases the ONT’s default compression level from 1 to 10. Even though its lossless, deep lossless, and raw methods reduce the storage size by 25\%, 44\%, and 88\%, respectively, it does not attempt any novel methods for compression.

Bzip2, often used for compressing large datasets, achieves better compression than  LZ77/LZ78-based compressors, reaching the performance of PPM (prediction by partial matching) family compressors. This library is implemented using the Burrows-Wheeler block sorting text compression algorithm and Huffman coding.

Gzip and Bzip2 utilise the Zlib compression library, which was developed in 1990. It was optimised for scalar, vector, and parallel computing on supercomputers like Cray computers, and the Zlib nerve system, including One-Step Index Pointers, was crucial for optimisation. Overall the Zlib Fortran version consisted of more than 200 dynamically usable subroutines for TPSA (Truncated Power Series Analysis) and Lie algebraic mapping analysis.

Huffman coding is a lossless compression algorithm introduced in 1952 that uses the least amount of symbols needed when coding a message minimising redundancy and reducing the number of digits needed to represent a message.

The run-length encoding system, another lossless compression method introduced for bandwidth or data-rate compression of television signals, has parameters that can be adjusted to find an optimum run-length encoding. The use of buffer stores of modest capacity is indicative of a potential compression strategy. The system is amenable to the depiction of both monochromatic as well as halftone images. The presence of a low-priced receiver in the system implies an economically viable mode of compression.

Arif and Anand are focused on the application of a run-length encoding (RLE) scheme to speech signals in various languages. The speech signals utilized are obtained from the IPA database, with particular emphasis on the German, American-English, and Japanese categories. At the encoder end, the RLE scheme is implemented to compress the speech signals, while at the decoder end, the compressed signal is reconstructed. To assess the efficiency of the RLE encoding algorithmic scheme, a comparison of compression ratios (CRs) is conducted for the three languages.

The current state-of-the-art nanopore signal data compression method is VBZ compression, which uses variable byte integer encoding. VBZ successively applies differential coding, zig-zag encoding StreamVByte, and zstd for lossless compression of nanopore raw signal data. Differential coding replaces the original data by their differences (deltas) between the successive data points. It is beneficial when the deltas are smaller than the actual data or more repetitive. In zig-zag encoding, negative and positive integers are interleaved together such as 0 maps to 0, -1 maps to 1, -2 maps to 2, etc. These encoding methods allow downstream compression methods to ignore the sign bit from the 2’s complement representation of the integers, preserving the same relative magnitudes of the numbers. After obtaining the zig-zag deltas by applying the above encoding methods, the svb (StreamVByte) integer compression technique is applied followed by the zstd algorithm. As the internal implementations of svb and zstd are highly optimized, the VBZ compression is significantly fast both in compression and decompression. It takes at least four runs throughout the dataset to apply the VBZ compression: one for differential coding, one for zig-zag encoding, one for svb, and at least one for zstd. The exact number of runs is needed for decompression as well. This compression method is a generic approach to every dataset. Therefore, analysing the data and exploiting its characteristics might lead to better compression mechanisms.

Dstall-fz combines VBBE21 with dynamic stall encoding and range coding modeled by order-0 order-1 secondary symbol estimation. It achieves a space-saving of 0.666. However, compressions and decompression take around 30.99 and 16.98 hours per TiB, respectively. 

Dstall-fz-1500 achieves an almost similar compression ratio in half the time. This strategy approximates the dstall-fz decision boundary and should be faster during compression. Whereas Dstall-fz has to encode a read two times because dstall must compare compressing the stall separately do not, so it must check both ways before proceeding. 

There exists an alternative approach that solely encodes the distinctions between a genome sequence and a reference sequence by utilising a variety of entropy coding techniques, resulting in noteworthy compression rates for genomic sequence data. The authors have effectively demonstrated the application of this approach utilising human mitochondrial genome sequences with high variability to serve as a testbed, achieving an impressive 345-fold compression rate with only a partial level of optimisation. However, a limitation of this approach is that the reference genome sequence should always be there to decode the signal.

Among the state-of-the-art algorithms for compressing nanopore sequencing data, dstall-fz has exhibited the highest compression ratio of 2.991729, albeit at the cost of being the slowest in terms of compression and decompression. Conversely, zstd-svb-zd is the fastest state-of-the-art algorithm, boasting a compression ratio of 2.928430 and a compression time of 1.04458386 hours per TiB. It is worth noting, however, that dstall-fz necessitates a compression time of 30.99850195 hours per TiB.

While the preceding section expounds upon the methodology of compressing prior to base-calling, additional areas of interest on compression post-base-calling. Meng et al.Meng have found a compression method for nanopore signals that employs approximate assembly. The authors have utilised MinHash indexing to efficiently index the reads, thereby facilitating the lookup of reads that overlap with any designated sequence. The authors have introduced NanoSpring, which focuses exclusively on the base sequences present in the FASTQ file, utilizing a mere 0.35-0.65 bits per base. This represents a 3-6× reduction compared to general-purpose compressors, such as gzip. Additionally, the speed of NanoSpring is significantly greater than that of the state-of-the-art tool, CoLoRd, especially when multiple threads are employed. They claim that it is four times faster in decompressing when 20 threads are utilized. Additional lossless techniques that are available subsequent to base calling include RENano and Enano.

Both lossless and lossy compression techniques have previously been examined. But the compression ratios of the present state-of-the-art are still not deemed satisfactory, and advancements are highly desirable. In lossy compression, the main focus is to identify and eliminate noise, and while this method has the potential to significantly decrease space requirements, there is a tradeoff in terms of accuracy. The success of this technique is evaluated by identity scores with the reference genome. Our research endeavors to examine a multitude of compression techniques as well as explore combinations of preexisting compression methods in order to achieve optimal compression ratios while ensuring that accuracy is not compromised.

## Methodology

## Experiment Setup and Implementation

## Results and Analysis

## Conclusion

## Publications
[//]: # "Note: Uncomment each once you uploaded the files to the repository"

1. [Project Proposal Presentation](https://docs.google.com/presentation/d/1B5NHxaCxq4HWcRp3xfms52uuq7upGqzzS-gKsycfYFk/edit?usp=sharing)
2. [Progress Presentation](https://docs.google.com/presentation/d/1SknsVYHWG2FfJuG1bv61YaqGDOdCnZuaPIiq5vtCEjM/edit?usp=sharing)
3. [Final Presentation](https://docs.google.com/presentation/d/1cpSIqvn2VKrk00jgBUI-Jm8HtgLYtgKnLqsMLzxlVk4/edit?usp=sharing)
<!-- 5. Author 1, Author 2 and Author 3 "Research paper title" (2021). [PDF](./). -->


## Links

[//]: # ( NOTE: EDIT THIS LINKS WITH YOUR REPO DETAILS )

- [Project Repository](https://github.com/cepdnaclk/e17-4yp-novel-compression-methods-for-nanopore-raw-signal-data)
- [Project Page](https://cepdnaclk.github.io/e17-4yp-novel-compression-methods-for-nanopore-raw-signal-data/)
- [Department of Computer Engineering](http://www.ce.pdn.ac.lk/)
- [University of Peradeniya](https://eng.pdn.ac.lk/)

[//]: # "Please refer this to learn more about Markdown syntax"
[//]: # "https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet"
